src/Compilers/CSharp/Test/Emit/Emit/CompilationEmitTests.cs

```diff
@@ -236,6 +236,122 @@ public static void Main()
+
+            var source2 = sourceTemplate.Replace("CHANGE", change2);
+            Compilation comp2 = CreateCompilationWithMscorlib(Parse(source2),
+                options: TestOptions.DebugDll.WithDeterministic(true), assemblyName: name);
+
+            byte[] image2;
+
+            using (var output = new MemoryStream())
+            using (var pdbOutput = new MemoryStream())
+            {
+                var emitResult = comp2.Emit(output, pdbOutput, options: EmitOptions.Default.WithEmitMetadataOnly(true));
+                Assert.True(emitResult.Success);
+                emitResult.Diagnostics.Verify();
+                image2 = output.ToArray();
+            }
+
+            Assert.True(AssertEx.SequenceEqual(image1, image2) == expectMatch);
+        }
+
+        [Theory,
+            InlineData("public int M() { error(); }", true),
+            InlineData("public int M() { error() }", false), // Should be true. See follow-up issue https://github.com/dotnet/roslyn/issues/17612
+            InlineData("public Error M() { return null; }", false), // Should be true. See follow-up issue https://github.com/dotnet/roslyn/issues/17612
+            ]
+        public void RefAssemblyDiagnostics(string change, bool expectSuccess)
+        {
+            string sourceTemplate = @"
+public class C
+{
+    CHANGE
+}
+";
+            string source = sourceTemplate.Replace("CHANGE", change);
+            string name = GetUniqueName();
+            CSharpCompilation comp1 = CreateCompilationWithMscorlib(Parse(source),
+                options: TestOptions.DebugDll.WithDeterministic(true), assemblyName: name);
+
+            byte[] image;
+
+            using (var output = new MemoryStream())
+            using (var pdbOutput = new MemoryStream())
+            {
+                var emitResult = comp1.Emit(output, pdbOutput, options: EmitOptions.Default.WithEmitMetadataOnly(true));
+                Assert.Equal(expectSuccess, emitResult.Success);
+                Assert.Equal(!expectSuccess, emitResult.Diagnostics.Any());
+                image = output.ToArray();
+            }
+        }
+
+        [Fact]
+        public void VerifyRefAssembly()
+        {
+            string source = @"
+public class PublicClass
+{
+    public void PublicMethod() { }
+    private void PrivateMethod() { }
+    protected void ProtectedMethod() { }
+    internal void InternalMethod() { }
+}
+";
+            CSharpCompilation comp = CreateCompilation(source, references: new[] { MscorlibRef },
+                options: TestOptions.DebugDll.WithDeterministic(true));
+
+            var emitRefOnly = EmitOptions.Default.WithEmitMetadataOnly(true);
+
+            var verifier = CompileAndVerify(comp, emitOptions: emitRefOnly, verify: false);
+            Assert.True(verifier.TestData.Methods.IsEmpty); // no method bodies
+
+            var image = comp.EmitToImageReference(emitRefOnly);
+            var comp2 = CreateCompilation("", references: new[] { MscorlibRef, image },
```

Comment:
>var comp2 = CreateCompilation("", references: new[] { MscorlibRef, image }, [](start = 12, length = 75)

I believe ```CompileAndVerify``` has built in service exactly for this purpose - it takes a delegate that verifies symbols of the produced assembly.